Add leakage noise in NoiseModel

a-corni opened the pull request:
Add the option to include leakage in NoiseModel.
It's currently done by modifying the parameter `with_leakage` in NoiseModel.
`with_leakage` is a boolean. It is associated with the noise type "leakage". 
For backward compatibility reason, with_leakage is an option in the schema of noise_model.
I have also added it with_leakage in the default values (if someone uses noise_types). However, I am not sure if the behaviour is correct: 
- by default, the parameter with_leakage is False in the __init__, and to have leakage the user should set `with_leakage` to True.
- however, if the user asks for leakage in noise_types, then I guess he doesn't want to also set with_leakage to True, so the default value of with_leakage in _LEGACY_DEFAULTS is True.
- In SimConfig, the user only defines leakage in the noises, and not via with_leakage, that is only a property instead of an argument.

HGSilveri commented:
Thanks for this @a-corni ! I was wondering, isn't there some sort of rate for leakage?

HGSilveri commented on pulser-core/pulser/noise_model.py:
```diff
@@ -366,9 +416,11 @@ def _check_eff_noise(
-            if operator.shape != (2, 2):
-                raise NotImplementedError(
-                    f"Operator's shape must be (2,2) not {operator.shape}."
+            if operator.shape not in possible_shapes:
+                raise ValueError(
+                    f"With{'' if with_leakage else 'out'} leakage, operator's "
+                    f"shape must be {possible_shapes[0]} or "
+                    f"{possible_shapes[1]}, not {operator.shape}."
```
So the changes for the leakage will also allow qutrit operators without leakage, right? I guess this will make a weird in-between state where our NoiseModel supports effective noise on qutrits but our `QutipEmulator` doesn't

a-corni replied:
You are right, I am adding the leakage in effective noise operators in a separate branch from this PR and I think I should modify this test there.

a-corni commented on pulser-core/pulser/noise_model.py:
```diff
@@ -312,8 +334,29 @@ def _find_relevant_params(
         # Disregard laser_waist when not defined
         if laser_waist is None:
             relevant_params.discard("laser_waist")
+        # Diregard leakage if with_leakage is False
+        if not with_leakage:
+            relevant_params.discard("with_leakage")
```
That's not needed indeed, if with_leakage is false then leakage will not be in true_noise_types at its creation :)

a-corni commented on pulser-core/pulser/noise_model.py:
```diff
@@ -353,6 +397,12 @@ def _check_eff_noise(
 
         if np.any(np.array(eff_noise_rates) < 0):
             raise ValueError("The provided rates must be greater than 0.")
+        print(with_leakage)
```
Yeah sorry I definitely was not focused in my last commit ^^'

a-corni commented:
> Thanks for this @a-corni ! ~I was wondering, isn't there some sort of rate for leakage?~ EDIT: Nevermind, I just remembered that it has to be given by the user through the effective noise parameters

Indeed ! I would like to first implement it for effective noise operator, and then have a look at implementing it with dephasing and depolarizing noise. For dephasing, we might want to have a error_dephasing_rate (like the hyperfine_dephasing_rate), but let's see this when it will be needed :) 

One thing I am not sure of is the way to go from this PR:
- should I merge it ? But then users can only define effective_noise_opers of shape (2, 2) and won't perform simulations with noise. Should I then add a NotImplementedError in Hamiltonian/QutipEmulator saying that simulation with leakage is not implemented yet ?
- should I change this PR to also incorporate the implementation of leakage in pulser-simulation ? Should I make a separate PR starting from these changes ?
I think the best is going with NotImplementedError in Hamiltonian for the moment... What is your opinion on this ?

HGSilveri commented:
> One thing I am not sure of is the way to go from this PR:
> 
>     * should I merge it ? But then users can only define effective_noise_opers of shape (2, 2) and won't perform simulations with noise. Should I then add a NotImplementedError in Hamiltonian/QutipEmulator saying that simulation with leakage is not implemented yet ?
> 
>     * should I change this PR to also incorporate the implementation of leakage in pulser-simulation ? Should I make a separate PR starting from these changes ?
>       I think the best is going with NotImplementedError in Hamiltonian for the moment... What is your opinion on this ?

Let's go with the `NotImplementedError`

HGSilveri commented on pulser-core/pulser/noise_model.py:
```diff
@@ -246,26 +244,26 @@ def to_tuple(obj: tuple) -> tuple:
             for nt_ in noise_types:
+                if nt_ == "leakage":
+                    raise ValueError(
+                        "'Leakage' cannot be explicitely defined in the noise"
```
nit:
```suggestion
                        "'leakage' cannot be explicitely defined in the noise"
```

HGSilveri commented on pulser-core/pulser/noise_model.py:
```diff
@@ -298,7 +296,11 @@ def to_tuple(obj: tuple) -> tuple:
             p: param_vals[p]
             for p in param_vals
-            if param_vals[p] is not None or (p in relevant_params)
+            if not (
+                param_vals[p] is None
+                or (isinstance(param_vals[p], bool) and not param_vals[p])
+            )
+            or p in relevant_params
```
I'm not sure, but maybe we can get away with
```suggestion
            if param_vals[p] or p in relevant_params
```
Did you try this?

a-corni replied:
I tried this, you have a misbehaviour with runs=0 (the value of run is not validated, and the NoiseModel is generated with runs=0)

HGSilveri replied:
Ok, in that case I'd say this special case is not worth it, I'd suggest reverting to the less strict validation that a `bool` is a boolean, not necessarily `True`. In reality, we also allow other parameters to be zero (when not strictly positive), so it's more consistent behaviour

a-corni replied:
I agree, I had the feeling of duplicating the test.

HGSilveri commented on pulser-core/pulser/noise_model.py:
```diff
@@ -416,11 +409,9 @@ def _check_eff_noise(
-                    f"{possible_shapes[1]}, not {operator.shape}."
+            if operator.shape != (2, 2):
+                raise NotImplementedError(
+                    f"Operator's shape must be (2,2) not {operator.shape}."
```
I think you can still allow (3, 3) if `with_leakage=True` (it will then hit the `NotImplementedError` on the hamiltonian but that's a different story), just not (4,4)

a-corni replied:
Yeah I agree,  I will do it this way :)

a-corni replied:
I have decided to go for something a bit more clever: raise an error depending on if the shape might be implemented some day or not (assuming that we might never support shapes outside (2, 2), (3, 3) without leakage, and (3, 3), (4, 4) with leakage.

HGSilveri commented on pulser-core/pulser/noise_model.py:
```diff
@@ -299,6 +322,7 @@ def _find_relevant_params(
         state_prep_error: float,
         amp_sigma: float,
         laser_waist: float | None,
+        with_leakage: bool,
```
You should remove this now

HGSilveri reviewed (commented):
I only have a few nits, so it's pretty much good for me. One thing though: wouldn't it perhaps make more sense to merge #715 first?

a-corni commented on tests/test_noise_model.py:
```diff
@@ -176,8 +176,28 @@ def matrices(self):
+        matrices["I4"] = np.eye(4)
         return matrices
 
+    @pytest.mark.parametrize("value", [False, True, 0, 1, 0.1])
+    def test_init_bool_like(self, value, matrices):
+        if isinstance(value, bool):
+            noise_model = NoiseModel(
+                eff_noise_rates=[0.1],
+                eff_noise_opers=[matrices["I3"] if value else matrices["I"]],
+                with_leakage=value,
+            )
+            assert noise_model.with_leakage == value
+            return
+        with pytest.raises(
+            ValueError, match=f"'with_leakage' must be a boolean, not {value}"
+        ):
+            noise_model = NoiseModel(
+                eff_noise_rates=[0.1],
+                eff_noise_opers=[matrices["I3"] if value else matrices["I"]],
+                with_leakage=value,
+            )
```
I can do this. Initially I inspired myself from the aforedefined test_init_prob_like: https://github.com/pasqal-io/Pulser/blob/add_leakage/tests/test_noise_model.py#L145

a-corni commented on tests/test_noise_model.py:
```diff
@@ -235,7 +289,12 @@ def test_relevant_params(self):
             "p_false_neg",
         }
-        assert NoiseModel._find_relevant_params({"SPAM"}, 0.1, 0.5, 100) == {
+        assert NoiseModel._find_relevant_params(
```
Oops sorry, thanks for the remark :)

a-corni commented:
> I only have a few nits, so it's pretty much good for me. One thing though: wouldn't it perhaps make more sense to merge #715 first?

Thanks for the review ! I have addressed the nits. As you want, we can definitely merge #715 before merging this PR 😄

HGSilveri commented:
> > I only have a few nits, so it's pretty much good for me. One thing though: wouldn't it perhaps make more sense to merge #715 first?
> 
> Thanks for the review ! I have addressed the nits. As you want, we can definitely merge #715 before merging this PR 😄

I thought that there were some `TODOs` here that could be done already, that's why I suggested it

HGSilveri reviewed (approved):
LGTM!

a-corni commented:
> > > I only have a few nits, so it's pretty much good for me. One thing though: wouldn't it perhaps make more sense to merge #715 first?
> > 
> > 
> > Thanks for the review ! I have addressed the nits. As you want, we can definitely merge #715 before merging this PR 😄
> 
> I thought that there were some `TODOs` here that could be done already, that's why I suggested it

That's True, I will let you merge your PR, I will delete the TODOs in this PR afterwards.

HGSilveri reviewed (approved):
Great!

HGSilveri commented on pulser-core/pulser/json/abstract_repr/schemas/noise-schema.json:
```diff
@@ -125,13 +128,13 @@
     },
     "NoiseType": {
       "enum": [
+        "leakage",
```
This was already there

HGSilveri commented on pulser-core/pulser/json/abstract_repr/schemas/noise-schema.json:
```diff
@@ -103,6 +103,9 @@
         },
         "temperature": {
           "type": "number"
+        },
+        "with_leakage": {
+          "type": "boolean"
```
I guess this is not needed, we can just check if `"leakage"` is in the `noise_types`. That way we won't need to make any modifications to the schema

HGSilveri reviewed (changes requested):
Sorry, I forgot to check the schema

a-corni reviewed (commented):
Thanks for this review of the schema. I have followed your suggestions. Should I keep 'with_leakage' property in the __repr__ as well ?

HGSilveri commented:
> Thanks for this review of the schema. I have followed your suggestions. Should I keep 'with_leakage' property in the **repr** as well ?

I thought about it and I'm torn actually. It's obviously redundant but if someone sets `with_leakage=True` I imagine they might expect to see it appear in the repr. I guess leaving it is the safer option, at least it won't cause any confusion

HGSilveri reviewed (approved):
Now I think we're good!

a-corni commented:
> > Thanks for this review of the schema. I have followed your suggestions. Should I keep 'with_leakage' property in the **repr** as well ?
> 
> I thought about it and I'm torn actually. It's obviously redundant but if someone sets `with_leakage=True` I imagine they might expect to see it appear in the repr. I guess leaving it is the safer option, at least it won't cause any confusion

That's also my feeling, let's go with this :)
